### DISCUSSION TITLE
Parsing apache access and error logs and inserting values in clickhouse

### DISCUSSION THREAD

**Discussion by SirUrban** [vectordotdev/vector#7772] [Q&A] [answered]
Hi,

since two weeks now I am struggling with the following task:
I have to receive apache access and error log as syslog messages and insert the separate values into clickhouse tables. With apache access logs I managed to do this using vector and a regex-parser transforms. With error logs I could not find out how to parse them correctly so that a clickhouse sink was able to insert the values into clickhouse.
Then I discovered the "parse_apache_log()" function and I thought it might be easier to use this instead of the regex-parser. But with this function I do have the problem that I do not understand how to get a result which can be handled by the clickhouse sink correctly.

In whole I'd like to do the following:
1. Define an input which receives the syslog messages. That's no problem. I did it like this:
```
[sources.in]
  address = "0.0.0.0:514" # required, required when mode = "tcp" or mode = "udp"
  mode = "tcp" # required
  type = "syslog" # required
```

2. I added a transforms section to modify some fields:
```
[transforms.field_modifier]
  type = "remap"
  inputs = ["in"]
  source = '''
  .uuid = uuid_v4()
  del(.facility)
  del(.severity)
  del(.source_type)
  del(.hostname)
  '''
```

3. I think I need to have a filter which separates access and error log message to be able to parse them differently (cause they do have different format):
```
[transforms.access_log_filter]
  type = "filter"
  inputs = ["field_modifier"]
  condition = '.appname == "apache_access"'

[transforms.error_log_filter]
  type = "filter"
  inputs = ["in"]
  condition = '.appname == "apache_error"'
```

4. I want to transform the syslog event and I want the following to be done:
- parse message field so that all parts (ip, status, path, length, user-agent ...) can be inserted by a sink into separate columns of a clickhouse table.
```
[transforms.access_log_parser]
  type = "remap"
  inputs = ["access_log_filter"]
  source = '''
  .message = parse_apache_log!(.message, "combined")
  '''
```

5. Last I have a sink to insert data into clickhouse:
```
[sinks.access_log]
  host = "http://localhost:8123" # required
  inputs = ["access_log_parser"] # required
  table = "access_log"
  type = "clickhouse" # required
  auth.password = "*****************"
  auth.strategy = "basic"
  auth.user = "default"
  encoding.timestamp_format = "unix"
```

I also will need a second sink for error logs - but as the sink for access logs does not work yet I left this out yet.

All this does not work. I do get a sink error:
```
Jun 07 11:05:17.847 ERROR sink{component_kind="sink" component_name=access_log component_type=clickhouse}:request{request_id=0}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 400 Bad Request"
Jun 07 11:05:17.847 ERROR sink{component_kind="sink" component_name=access_log component_type=clickhouse}:request{request_id=0}: vector::sinks::util::sink: Response wasn't successful. response=Response { status: 400, version: HTTP/1.1, headers: {"date": "Mon, 07 Jun 2021 09:05:17 GMT", "connection": "Keep-Alive", "content-type": "text/tab-separated-values; charset=UTF-8", "x-clickhouse-server-display-name": "logexplorer.az-logexplorer.scalecommerce", "transfer-encoding": "chunked", "x-clickhouse-query-id": "aa5965b2-3980-4ab3-82cb-2626492d4c10", "x-clickhouse-format": "TabSeparated", "x-clickhouse-timezone": "Europe/Berlin", "x-clickhouse-exception-code": "26", "keep-alive": "timeout=3", "x-clickhouse-summary": "{\"read_rows\":\"0\",\"read_bytes\":\"0\",\"written_rows\":\"0\",\"written_bytes\":\"0\",\"total_rows_to_read\":\"0\"}"}, body: b"Code: 26, e.displayText() = DB::ParsingException: Cannot parse JSON string: expected opening quote: (while reading the value of key message): (at row 1)\n (version 21.5.5.12 (official build))\n" }
```

I assume that step 4 is wrong but I have no idea how to do it the right way. For the parsing part I also tried this one:
```
[transforms.access_log_parser]
  type = "remap"
  inputs = ["access_log_filter"]
  source = '''
  . = parse_apache_log!(.message, "combined")
  '''
```
This does work so far that values are inserted in clickhouse - but only the parsed parts of the message field not the contents of other fields (uuid, appname, source_ip).
So if anybody could help me to get the message parsed without loosing the other fields - and also how to do it with error logs I'd very much appreciate some support!

**Comment by spencergilbert** [comment, marked as answer]
I expect you'll want to [merge](https://vector.dev/docs/reference/vrl/functions/#merge) the parsed log event.

The first config you used set the `.message` field to the parsed results (an object), and the second overwrote the entire event ("`.`") with the parsed results.

So, I think you should either use the `merge` function, or an object merge assignment (example [here](https://vector.dev/docs/reference/vrl/expressions/#assignment)) like so:

```
parsed = parse_apache_log!(.message, "combined")
merge!(., parsed)
```

or

```
. |= parse_apache_log!(.message, "combined")
```

**Reply by SirUrban** [reply]
Hi @spencergilbert,

thx a lot - this helped! I managed to get the apache access log stuff done now. That's fine.
But I am still struggling with the error logs ... :-/ I've got two problems and hope you might be able to help once more?

First is regarding the timestamp from syslog message. The original apache error syslog message taken from rsylog output on the sending server looks like this:
```
msg 0: [Tue Jun 08 11:54:08.929301 2021] [php7:emerg] [pid 1374899] [client 95.223.77.60:41888] PHP Parse error:  syntax error, unexpec...
```

Same message at the receiving server - received by rsyslogd - looks like this:
```
msg 0: <131>Jun  8 11:54:08 master apache_error [Tue Jun 08 11:54:08.929301 2021] [php7:emerg] [pid 1374899] [client 95.223.77.60:41888...
```

If I receive the message via vector the fields object looks like this:
```
{ fields: {"appname": Bytes(b"apache_error"), "facility": Bytes(b"local0"), "host": Bytes(b"master"), "hostname": Bytes(b"master"), "message": Bytes(b"[php7:emerg] [pid 1374896] [client 95.223.77.60:59042] PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14"), "severity": Bytes(b"err"), "source_ip": Bytes(b"10.1.134.8"), "source_type": Bytes(b"syslog"), "timestamp": Timestamp(2021-06-08T09:57:41Z)}, metadata: EventMetadata }
```

As you can see the message field is missing the timestamp and this seems to be a problem for the `parse_apache_log` function. It gives me an error without the timestamp - but works if I add it manually:

```
$ parse_apache_log!(s'[php7:emerg] [pid 4803:tid 4803] [client 95.223.77.60:35106] PHP Parse error:  syntax error, unexpected \'->\' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14', timestamp_format: "%Y-%m-%d %H:%M:%S.%f", format: "error")
function call error for "parse_apache_log" at (0:313): failed parsing common log line

$ parse_apache_log!(s'[2021-06-04 15:40:27.138633] [php7:emerg] [pid 4803:tid 4803] [client 95.223.77.60:35106] PHP Parse error:  syntax error, unexpected \'->\' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14', timestamp_format: "%Y-%m-%d %H:%M:%S.%f", format: "error")
{ "client": "95.223.77.60", "message": "PHP Parse error:  syntax error, unexpected \'->\' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14", "module": "php7", "pid": 4803, "port": 35106, "severity": "emerg", "thread": "4803", "timestamp": t'2021-06-04T13:40:27.000138633Z' }
```

It doesn't matter if I also leave the timestamp_format param out, it does not work anyway. So why is the timestamp missing in vector fields? Is there any way to get it working without the timestamp? or to add the timestamp by some transforms magic?

Another problem I have is the `[pid 4803:tid 4803]` part of the message. On our servers the standard apache ErrorLogFormat does not contain the "tid"-Part. It's only `[pid 4803]`. But the parse does not work if there's no "tid". I was able to fix this by setting up an individual ErrorLogFormat on sending servers - but this would be a quite uncomfortable solution as I would have to change this on quite a lot servers ... - So, any idea about this issue?

**Reply by spencergilbert** [reply]
```
{ fields: {"appname": Bytes(b"apache_error"), "facility": Bytes(b"local0"), "host": Bytes(b"master"), "hostname": Bytes(b"master"), "message": Bytes(b"[php7:emerg] [pid 1374896] [client 95.223.77.60:59042] PHP Parse error: syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14"), "severity": Bytes(b"err"), "source_ip": Bytes(b"10.1.134.8"), "source_type": Bytes(b"syslog"), "timestamp": Timestamp(2021-06-08T09:57:41Z)}, metadata: EventMetadata }
```

As the message is getting pulled in by the `syslog` source it's getting parsed internally, see [output](https://vector.dev/docs/reference/configuration/sources/syslog/#output) and [examples](https://vector.dev/docs/reference/configuration/sources/syslog/#examples)

I'll try and think (and ask around) of a good solution here.

**Reply by jszwedko** [reply]
Hmm. There appears to be a bug here with the syslog parser. Given the example RFC 3164 message:

```
<131>Jun 8 11:54:08 master apache_error [Tue Jun 08 11:54:08.929301 2021] [php7:emerg] [pid 1374899] [client 95.223.77.60:41888...
```

The first timestamp is the syslog timestamp, and the one that Vector will automatically parse out. The second timestamp is part of the message though, and should still appear in `.message`. I can verify that it is getting dropped though. I think Vector is trying to interpret it as the process id, which it shouldn't be.  I opened https://github.com/timberio/vector/issues/7796 to track this.



> Another problem I have is the `[pid 4803:tid 4803]` part of the message. On our servers the standard apache ErrorLogFormat does not contain the "tid"-Part. It's only `[pid 4803]`. But the parse does not work if there's no "tid". I was able to fix this by setting up an individual ErrorLogFormat on sending servers - but this would be a quite uncomfortable solution as I would have to change this on quite a lot servers ... - So, any idea about this issue?

I'm curious what version of apache you are using. It seems like we should be supporting messages without the `tid` part if you are seeing that with the standard `ErrorLogFormat`. To confirm, that is unmodified, yes?

**Reply by SirUrban** [reply]
Our apache version:
```
Server version: Apache/2.4.29 (Ubuntu)
Server built:   2020-08-12T21:33:25
Server's Module Magic Number: 20120211:68
Server loaded:  APR 1.6.3, APR-UTIL 1.6.1
Compiled using: APR 1.6.3, APR-UTIL 1.6.1
Architecture:   64-bit
Server MPM:     prefork
  threaded:     no
    forked:     yes (variable process count)
Server compiled with....
 -D APR_HAS_SENDFILE
 -D APR_HAS_MMAP
 -D APR_HAVE_IPV6 (IPv4-mapped addresses enabled)
 -D APR_USE_SYSVSEM_SERIALIZE
 -D APR_USE_PTHREAD_SERIALIZE
 -D SINGLE_LISTEN_UNSERIALIZED_ACCEPT
 -D APR_HAS_OTHER_CHILD
 -D AP_HAVE_RELIABLE_PIPED_LOGS
 -D DYNAMIC_MODULE_LIMIT=256
 -D HTTPD_ROOT="/etc/apache2"
 -D SUEXEC_BIN="/usr/lib/apache2/suexec"
 -D DEFAULT_PIDLOG="/var/run/apache2.pid"
 -D DEFAULT_SCOREBOARD="logs/apache_runtime_status"
 -D DEFAULT_ERRORLOG="logs/error_log"
 -D AP_TYPES_CONFIG_FILE="mime.types"
 -D SERVER_CONFIG_FILE="apache2.conf"
 ```

There's no ErrorLogFormat directive set, so default is used. The fact we do have no tid may be connected with our special environment: we do use docker containers like VMs. So a complete ubuntu runs inside a docker container and apache process is managed by supervisord. May be that inside the docker container we do not have a thread id?

**Reply by jszwedko** [reply]
Thanks for confirming @SirUrban ! I opened https://github.com/timberio/vector/issues/7827 to track the failed parsing when `tid` is missing. I believe we can get to both of those issues soon.

**Reply by SirUrban** [reply]
BTW: Regarding the timestamp issue - same thing seems to happen with nginx logs.

Original syslog message:
```
ruleset.c: processBATCH: next msg 0: [11/Jun/2021:09:12:10 +0200] az-logexplorer.scalecommerce.cloud 95.223.77.60 200 200 9675 0.000 0.072 0.073 GET / HTTP/1.1 ...
```

message in vector:
```
"message": Bytes(b"az-logexplorer.scalecommerce.cloud 95.223.77.60 200 200 9675 0.000 0.072 0.073 GET / HTTP/1.1"), "severity": Bytes(b"info")
```
Again the timestamp is missing.

**Reply by SirUrban** [reply]
May be there's a general problem with parsing timestamps? I tried to use the key-value-parser and got this:
```
$ parse_key_value!("time=\"2021.2021-06-11 09:54:11Z\", severity=[php7:emerg] , process=1700689 , message=\"PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14\"")
{ ", message": "PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14", ", process": "1700689", "09:54:11Z", severity": "[php7:emerg]", "time": "\"2021.2021-06-11" }
```
Looks a bit strange, doesn't it?

**Reply by jszwedko** [reply]
Yeah, I think that is the same issue as https://github.com/timberio/vector/issues/7796 regarding it incorrectly consuming the timestamp.

**Reply by jszwedko** [reply]
Hmm, yeah, regarding the key value parser, it seems like it isn't picking up the quotes around the timestamp as delimiting the value. I'll open a separate issue for that. It is supposed to (and we can see it doing so with `message`).
